### PR TITLE
Trigger an orchestration on any crd update

### PR DESCRIPTION
**Pull request by travisn**
Signed-off-by: travisn <tnielsen@redhat.com>

**Description of your changes:**
The orchestrations need to be triggered on any update to the crd. Users today restart the operator for various scenarios when the operator could take care of it automatically. If unsupported options are updated, they should just be ignored.

The subsequent change needed for this is it became apparent there was a race condition for updating the mons. There are two goroutines that are working with the mons. The first is an orchestration that is triggered by the operator at startup, or when the cluster crd is updated. The second is the health check that triggers periodically by default every 45 seconds. These goroutines must not try to make updates at the same time. A mutex is added so one will block if the other is still active. The mutex is only active for the duration of working with mons, and not an entire orchestration of mgr, osd, etc

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

### CODE REVIEW

File: pkg/operator/ceph/cluster/cluster.go [hunk 1]
```diff
@@ -72,7 +72,8 @@ func newCluster(c *cephv1.CephCluster, context *clusterd.Context) *cluster {
 		Spec:      &c.Spec,
 		context:   context,
 		stopCh:    make(chan struct{}),
-		ownerRef:  ClusterOwnerRef(c.Namespace, string(c.UID)),
+		ownerRef:  ownerRef,
+		mons:      mon.New(context, c.Namespace, c.Spec.DataDirHostPath, c.Spec.Network.HostNetwork, ownerRef),
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/cluster.go]: Is there a way we can start reducing the number of parameters we are passing around? I feel like I spend a lot of time changing function definitions and calls. Why don't we choose to pass around the cluster spec more often so the daemon operator code paths could have access to all the information they might need?

File: pkg/operator/ceph/cluster/cluster.go [hunk 2]
```diff
@@ -63,6 +62,7 @@ type cluster struct {
 }
 
 func newCluster(c *cephv1.CephCluster, context *clusterd.Context) *cluster {
+	ownerRef := ClusterOwnerRef(c.Namespace, string(c.UID))
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/cluster.go]: Not an issue with this PR's changes, but something odd I see that might merit changing, the `ClusterOwnerRef` function accepts a string UID, and then it converts it again to a UID. Would it be better for this function to accept `type.UID`?

```go
func ClusterOwnerRef(namespace, clusterID string) metav1.OwnerReference {
	blockOwner := true
	return metav1.OwnerReference{
		APIVersion:         ClusterResource.Version,
		Kind:               ClusterResource.Kind,
		Name:               namespace,
		UID:                types.UID(clusterID),
		BlockOwnerDeletion: &blockOwner,
	}
}
```

File: pkg/operator/ceph/cluster/cluster.go [hunk 3]
```diff
@@ -261,78 +261,18 @@ func (c *cluster) createInitialCrushMap() error {
...
+	sort.Sort(rookv1alpha2.NodesByName(oldCluster.Storage.Nodes))
+	sort.Sort(rookv1alpha2.NodesByName(newCluster.Storage.Nodes))
 
-	if oldCluster.Mon.Count != newCluster.Mon.Count {
-		logger.Infof("number of mons have changed from %d to %d. The health check will update the mons...", oldCluster.Mon.Count, newCluster.Mon.Count)
-		clusterRef.mons.MonCountMutex.Lock()
-		clusterRef.mons.Count = newCluster.Mon.Count
-		clusterRef.mons.MonCountMutex.Unlock()
+	// any change in the crd will trigger an orchestration
+	if !reflect.DeepEqual(oldCluster, newCluster) {
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/cluster.go]: For my own clarity, this `DeepEqual` is part of the generated code, yes?

File: pkg/operator/ceph/cluster/mon/mon.go [hunk 4]
```diff
@@ -85,7 +85,7 @@ type Cluster struct {
 	cephVersion          cephv1.CephVersionSpec
 	Count                int
 	AllowMultiplePerNode bool
-	MonCountMutex        sync.Mutex
+	monMutex             sync.Mutex
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/mon/mon.go]: Is it beneficial to call this `monMutex` over just `mutex`?

File: pkg/operator/ceph/cluster/mon/mon.go [hunk 5]
```diff
@@ -129,44 +129,43 @@ type NodeInfo struct {
...
+func (c *Cluster) Start(clusterInfo *cephconfig.ClusterInfo, rookVersion string, cephVersion cephv1.CephVersionSpec,
+	mon cephv1.MonSpec, placement rookalpha.Placement, resources v1.ResourceRequirements) (*cephconfig.ClusterInfo, error) {
+
+	// Only one goroutine can orchestrate the mons at a time
+	logger.Infof("Acquiring lock for mon orchestration")
+	c.monMutex.Lock()
+	defer c.monMutex.Unlock()
+	logger.Infof("Acquired lock for mon orchestration")
+
+	c.clusterInfo = clusterInfo
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/mon/mon.go]: Here, there is a lot of boilerplate to copy various cluster CRD items into the mon's `Cluster`. Is there a reason not to just use the cluster spec?

File: pkg/operator/ceph/cluster/mon/mon_test.go [hunk 6]
```diff
@@ -123,6 +123,19 @@ func newCluster(context *clusterd.Context, namespace string, hostNetwork bool, a
 	}
 }
 
+// testStart will orchestrate the mons with the existing settings on the mon cluster
+func testStart(c *Cluster) (*cephconfig.ClusterInfo, error) {
+	return c.Start(c.clusterInfo, c.rookVersion, c.cephVersion, cephv1.MonSpec{Count: c.Count, AllowMultiplePerNode: c.AllowMultiplePerNode},
+		rookalpha.Placement{}, v1.ResourceRequirements{})
+}
+
+func setTestMonSettings(c *Cluster, currentMons int, mon cephv1.MonSpec, rookVersion string) {
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/mon/mon_test.go]: I'm not super clear on what the purpose of this is. Maybe a comment or rename would help?

File: pkg/operator/ceph/cluster/cluster.go (hunk 2 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/cluster.go]: agreed, we should use the UID type instead

File: pkg/operator/ceph/cluster/cluster.go (hunk 1 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/cluster.go]: It's a great question. I'll look at this again and consider just passing the ClusterSpec type and reduce the params needed. It's a fine line between making it clear what is expected in a function call and passing structs with more properties than are needed.

File: pkg/operator/ceph/cluster/cluster.go (hunk 3 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/cluster.go]: it's actually on the `reflect` package. So if any fields change in the struct, we will notice the difference. Maybe you're thinking of the `DeepCopy()` on the generated code, I don't see a `DeepEqual()` there

File: pkg/operator/ceph/cluster/mon/mon.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/mon/mon.go]: How about `orchestrationMutex`? `mutex` would make it too easy to use for an unrelated purpose in the future.

File: pkg/operator/ceph/cluster/mon/mon.go (hunk 5 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/mon/mon.go]: Agreed, when i'm simplifying the params i'll look at reworking this.

File: pkg/operator/ceph/cluster/mon/mon_test.go (hunk 6 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/mon/mon_test.go]: this could be simplified by using a cluster spec as well, but if i keep it i'll at least add a comment

File: pkg/operator/ceph/cluster/cluster.go [hunk 7]
```diff
@@ -261,78 +261,18 @@ func (c *cluster) createInitialCrushMap() error {
...
+		logger.Infof("The Cluster CRD has changed")
```
Comment (leseb) [inline pkg/operator/ceph/cluster/cluster.go]: Perhaps it's worth mentioning what changed?

File: pkg/operator/ceph/cluster/mon/mon.go [hunk 8]
```diff
@@ -129,44 +129,43 @@ type NodeInfo struct {
...
 	}
 }
 
 // Start begins the process of running a cluster of Ceph mons.
-func (c *Cluster) Start() (*cephconfig.ClusterInfo, error) {
+func (c *Cluster) Start(clusterInfo *cephconfig.ClusterInfo, rookVersion string, cephVersion cephv1.CephVersionSpec,
+	mon cephv1.MonSpec, placement rookalpha.Placement, resources v1.ResourceRequirements) (*cephconfig.ClusterInfo, error) {
+
+	// Only one goroutine can orchestrate the mons at a time
+	logger.Infof("Acquiring lock for mon orchestration")
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/mon.go]:
```suggestion
	logger. Debugf("Acquiring lock for mon orchestration")
```

File: pkg/operator/ceph/cluster/mon/mon.go (hunk 5 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/mon/mon.go]: This is also used by `health.go`, perhaps a helper method makes sense? Later it could serve other purpose too.

File: pkg/operator/ceph/cluster/cluster.go (hunk 7 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/cluster.go]: Any suggestions on how to do that? It doesn't scale to analyze individual properties and it doesn't seem helpful to dump the entire old and new CRDs to the log.

File: pkg/operator/ceph/cluster/mon/mon.go (hunk 5 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/mon/mon.go]: @leseb What helper would make sense here? The health.go doesn't update these properties, it just consumes them.

File: pkg/operator/ceph/cluster/cluster.go (hunk 1 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/cluster.go]: Ok, pushed changes to use the cluster spec instead of individual args. While it does simplify the code, I'm still torn between passing minimal info that a package needs vs an object which contains a lot of extra properties.

File: pkg/operator/ceph/cluster/cluster.go (hunk 7 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/cluster.go]: Can't we print a diff or something? I'm not saying it has to be pretty, even raw output would be valuable. I suspect cloud admin will look at their logs and then they will have a hard time remembering what changed just by seeing "The Cluster CRD has changed".

File: pkg/operator/ceph/cluster/mon/mon.go (hunk 5 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/mon/mon.go]: @travisn I meant a helper method to acquire the lock.

Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/mon/mon.go]: I think you are talking about something like this, Seb?

```go
func (c *Cluster) acquireOrchestrationLock() {
	logger.Debugf("Acquiring lock for mon orchestration")
	c.orchestrationMutex.Lock()
	logger.Debugf("Acquired lock for mon orchestration")
}

func (c *Cluster) releaseOrchestrationLock() {
	c.orchestrationMutex.Unlock()
	logger.Debugf("Released lock for mon orchestration")
}
```

and then the 4 lines above become:
```go
c.acquireOrchestrationLock()
defer c.releaseOrchestrationLock()
```

It at least makes the debug messaging standard for whatever method is doing the locking/unlocking.

Comment (leseb) [reply pkg/operator/ceph/cluster/mon/mon.go]: Exactly 👍

File: pkg/operator/ceph/cluster/cluster.go (hunk 1 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/cluster.go]: I share some of your hesitation. Maybe a middle ground (for the future) would be to make a struct that contains global cluster config that any daemon might want to know (e.g., cluster context, rook and ceph versions, hostPath, namespace, owner ref to cluster object, etc.), and then the individual daemons would get that global spec as well as the spec specific to them (like filesystem spec for file/mdses, storage for osds).

File: pkg/operator/ceph/cluster/cluster.go [hunk 9]
```diff
@@ -286,79 +281,20 @@ func (c *cluster) createInitialCrushMap() error {
...
-
-	if oldCluster.CephVersion.Image != newCluster.CephVersion.Image {
-		logger.Infof("ceph version changing from %s to %s", oldCluster.CephVersion.Image, newCluster.CephVersion.Image)
-		changeFound = true
+	sort.Sort(rookv1alpha2.NodesByName(oldCluster.Storage.Nodes))
+	sort.Sort(rookv1alpha2.NodesByName(newCluster.Storage.Nodes))
+
+	// any change in the crd will trigger an orchestration
+	if !reflect.DeepEqual(oldCluster, newCluster) {
+		diff := cmp.Diff(oldCluster, newCluster)
```
Comment (galexrt) [inline pkg/operator/ceph/cluster/cluster.go]: Are we also reacting to `spec.status` updates?

Comment (travisn) [reply pkg/operator/ceph/cluster/cluster.go]: No, I'm only expecting to react to the core `spec` which defines the desired state. The `status` is actually not part of the spec.

File: pkg/operator/ceph/cluster/mon/mon.go [hunk 10]
```diff
@@ -129,47 +124,39 @@ type NodeInfo struct {
...
+		waitForStart:        true,
+		monPodRetryInterval: 6 * time.Second,
+		monPodTimeout:       5 * time.Minute,
+		monTimeoutList:      map[string]time.Time{},
+		HostNetwork:         hostNetwork,
 		mapping: &Mapping{
 			Node: map[string]*NodeInfo{},
 			Port: map[string]int32{},
 		},
-		resources: resources,
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/mon.go]: Why did you remove that? It's needed, I'm going to include it again in: https://github.com/rook/rook/pull/2764

Comment (travisn) [reply pkg/operator/ceph/cluster/mon/mon.go]: Instead of a `resources` var on the `Cluster` spec, now we can just access it with the `spec` directly from the `ClusterSpec` type. See the usage [here](https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/mon/spec.go#L153).
